### ISSUE TITLE
Fail to propagate by (bi)clauses

### ISSUE THREAD

**Issue by shnarazk** [shnarazk/splr#113]
## case 1

```text
$ splr aes.cnf
aes.cnf                                      501284,2928183 |time:    37.88
 #conflict:          0, #decision:        45110, #propagate:         219056
  Assignment|#rem:   330576, #ass:        0, #elm:   170708, prg%:  34.0541
      Clause|Remv:        0, LBD2:        0, Binc:  1071881, Perm:  1743759
     Restart|#BLK:        0, #RST:        0, trgr:        1, peak:        1
         LBD|avrg:   0.0000, trnd:      NaN, depG:   0.0000, /dpc:     0.00
    Conflict|tASG:      NaN, cLvl:     0.00, bLvl:     0.00, /ppc:     0.00
        misc|proc:        1, #sub:    79540, core:   501284, /cpr:     0.00
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `0`,
 right: `1`: found a strange biclause 66C
 -    -8L: Some(false) at level   0 by    Asserted at 0,  false
 -  1261L:        None at level   1 by     Not assigned,  false
', src/cdb/db.rs:1235:17
```

## case 2

```
$ splr aes.cnf
aes.cnf                                      501284,2928183 |time:    36.09
 #conflict:          0, #decision:        45115, #propagate:         219056
  Assignment|#rem:   330744, #ass:        0, #elm:   170540, prg%:  34.0206
      Clause|Remv:        0, LBD2:        0, Binc:  1072019, Perm:  1744273
     Restart|#BLK:        0, #RST:        0, trgr:        1, peak:        1
         LBD|avrg:   0.0000, trnd:      NaN, depG:   0.0000, /dpc:     0.00
    Conflict|tASG:      NaN, cLvl:     0.00, bLvl:     0.00, /ppc:     0.00
        misc|proc:        1, #sub:    79557, core:   501284, /cpr:     0.00
thread 'main' panicked at 'found a strange biclause 586961C
 -   64814L: Some(false) at level   1 by Implied by B3683537C, false
 -   62087L:        None at level   2 by         Not assigned, false
', src/cdb/db.rs:1235:17
```

## Reasons

1. `cdb.bi_clause` used wrong keys to store bi-clauses.
2. try to assert without backtack
3. TBU

**Comment by shnarazk** [comment]
``` splr aes.cnf
aes.cnf                                      501284,2928183 |time:   129.93
 #conflict:     167637, #decision:       832332, #propagate:       98349842
  Assignment|#rem:   330617, #ass:        0, #elm:   170667, prg%:  34.0460
      Clause|Remv:    70785, LBD2:     2900, Binc:  1031580, Perm:  1581442
     Restart|#BLK:      262, #RST:      552, trgr:        2, peak:       16
         LBD|avrg:   8.1575, trnd:   0.9389, depG:   3.3392, /dpc:     6.21
    Conflict|tASG:   0.2558, cLvl:   302.65, bLvl:   300.15, /ppc:  1178.74
        misc|proc:        7, #sub:    90162, core:        0, /cpr:   330.90
After extending the model
| on trail |   time | level |   literal  |  assignment |              reason | last move |
|   101655 |      0 |     0 |    206473  | Some(false) |Asserted at 0        |          0 |
|   106323 |      0 |     0 |    204789  | Some(false) |Asserted at 0        |          0 |
|   110048 |      0 |     0 |    205114  | Some(false) |Asserted at 0        |          0 |
|   113740 |      0 |     0 |    208563  | Some(false) |Asserted at 0        |          0 |
|   117355 |      0 |     0 |    204040  | Some(false) |Asserted at 0        |          0 |
|   120940 |      0 |     0 |    207015  | Some(false) |Asserted at 0        |          0 |
|   124455 |      0 |     0 |    204321  | Some(false) |Asserted at 0        |          0 |
|   127917 |      0 |     0 |    205388  | Some(false) |Asserted at 0        |          0 |
|   131335 |      0 |     0 |    206998  | Some(false) |Asserted at 0        |          0 |
|   134698 |      0 |     0 |    203880  | Some(false) |Asserted at 0        |          0 |
|   137980 |      0 |     0 |    208296  | Some(false) |Asserted at 0        |          0 |
|   141248 |      0 |     0 |    208636  | Some(false) |Asserted at 0        |          0 |
|   144440 |      0 |     0 |    204361  | Some(false) |Asserted at 0        |          0 |
|   147561 |      0 |     0 |    207069  | Some(false) |Asserted at 0        |          0 |
|   150570 |      0 |     0 |    203703  | Some(false) |Asserted at 0        |          0 |
|   153557 |      0 |     0 |    204996  | Some(false) |Asserted at 0        |          0 |
|   156478 |      0 |     0 |    208416  | Some(false) |Asserted at 0        |          0 |
|   159323 |      0 |     0 |    203720  | Some(false) |Asserted at 0        |          0 |
|   162112 |      0 |     0 |    204592  | Some(false) |Asserted at 0        |          0 |
|   164861 |      0 |     0 |    206979  | Some(false) |Asserted at 0        |          0 |
|   167523 |      0 |     0 |    203719  | Some(false) |Asserted at 0        |          0 |
|   170091 |      0 |     0 |    204393  | Some(false) |Asserted at 0        |          0 |
|   172504 |      0 |     0 |    204712  | Some(false) |Asserted at 0        |          0 |
|   174718 |      0 |     0 |    208177  | Some(false) |Asserted at 0        |          0 |
|   265601 | 167507 |    23 |    205441  | Some(false) |Implied by B3153809C |          0 |
|   265618 | 167507 |    23 |    203670  | Some(false) |Implied by 1271479C  |          0 |
|   265641 | 167507 |    23 |    205113  | Some(false) |Implied by B3822847C |          0 |
|   265646 | 167507 |    23 |    207297  | Some(false) |Implied by B3151927C |          0 |
|   265868 | 167507 |    23 |    204689  | Some(false) |Implied by B3146079C |          0 |
|   265871 | 167507 |    23 |    206271  | Some(false) |Implied by B3147049C |          0 |
|   265874 | 167507 |    23 |    207810* | Some(false) |Implied by B3153519C |          0 |
|   265875 | 167507 |    23 |    205140* | Some(false) |Implied by B3784906C |          0 |
|   312319 | 167605 |   337 |   -203234  | Some(false) |Implied by 3576936C  |          0 |
clause detail: {[207810, 205140, 203670, 203703, 203720, 203880, 204040, 204321, 204361, 203719, 204592, 204393, 204712, 204789, 204996, 206271, 205114, -203234, 205388, 205441, 205113, 206473, 206979, 206998, 207015, 207069, 207297, 204689, 208177, 208296, 208416, 208563, 208636]b0}
 - watch 207810L has watch cache [205140L]
 - watch 205140L has watch cache [207810L]
 which was born at 0
 which was used at 0
thread 'main' panicked at 'Before extending, NC 167637, Level 394 generated assignment(false) falsifies by 3783494C', src/solver/search.rs:399:9
```

**Comment by shnarazk** [comment]
Probably fixed by c25e8fa
